content/post/formatting-with-style.md

```diff
@@ -0,0 +1,247 @@
+when new code is pull-request-ed or pushed. Those analyzers are very
+easy to define (they are based on the gRPC toolsuite).
+
+In our case, when new code is pushed, we want to learn from a codebase
+to model its style as precisely as possible. When new code is
+pull-request-ed, we analyze it to detect any styling issue and report
+it for consideration during code review.
+
+## Unsupervised Learning
+
+Let's now dive into the Machine Learning side of things! In our first
+approach and to come up with a solution that completely addressed this
+problem of linting code given an existing codebase, we defined several
+keypoints to achieve:
+
+- only require the code of one repository to be train our model
+- output something interpretable (to be accountable and establish
+  trust with our users —a key point of Machine Learning in production)
+- come up with an correctable model: if a team is not happy with the
+  suggestions of the machine-learning based model, it should be able
+  to correct its suggestions
+
+To satisfy all those needs with a single approach, we turned to good
+old, battle-tested machine learning methods: language models and
+decision trees.
+
+### Language models
+
+To define the style of a repository, we summoned a tool used every day
+by Natural Language Processing practitioners: language models. A
+language model tries to find out which next word is the most likely
+given the previous words in a sentence. In our case, we'll use them to
+find out which next token is the most likely given the previous tokens
+in some code.
+
+Consider this example where we try to find out what `_` could be:
+
+    if a_
+
+Given the previous tokens `if`, ` ` and `a`, what would be the most
+likely next token `_`? There is not a single correct answer but rather
+a probability distribution over possible tokens. We could for example
+learn:
+
+    p(" ") = 0.4
+    p(":") = 0.6
+
+`p(" ")` would model the cases where the `if` condition is complex
+(`if a and len(some_list) > 5:`) and `p(":")` would model the cases
+where the `if` condition is already complete (`if a:`).
+
+A language model is simply the set of such probability distributions
+computed for all interesting inputs (all interesting sequences of
+previous tokens).
+
+This model is rather simple and has a big advantage: it doesn't
+require any labeled training data. That means that the engineers that
+will use the model won't have to spend time giving examples of correct
+style and bad style: we'll learn that without any intervention!
+
+To give some more power to our language models, we do not restrict the
+tokens we base our predictions on to previous tokens but also to the
+words that follow the character that we are predicting. Still, the
+idea is very close to a vanilla language model.
+
+Before diving into the specifics of the models we used, let's
+elaborate on this point and review how we create the window of tokens
+to consider to predict a formatting element and how we model the
```

Comment:
```suggestion
to consider and how we model the
```

Otherwise, "formatting element" is a new entity which we do not mention before (do we?) and readers can be confused